src/discordcr/voice.cr

```diff
@@ -0,0 +1,268 @@
+require "./mappings/gateway"
+require "./mappings/vws"
+require "./websocket"
+require "./sodium"
+
+module Discord
+  class VoiceClient
+    UDP_PROTOCOL = "udp"
+
+    # The mode that tells Discord we want to send encrypted audio
+    ENCRYPTED_MODE = "xsalsa20_poly1305"
+
+    OP_IDENTIFY            = 0
+    OP_SELECT_PROTOCOL     = 1
+    OP_READY               = 2
+    OP_HEARTBEAT           = 3
+    OP_SESSION_DESCRIPTION = 4
+    OP_SPEAKING            = 5
+    OP_HELLO               = 8
+
+    # The heartbeat is the same every time, so there's no point in constructing
+    # it from an object
+    HEARTBEAT_JSON = %({"op":#{OP_HEARTBEAT},"d":null})
+
+    @udp : VoiceUDP
+
+    @sequence : UInt16 = 0_u16
+    @time : UInt32 = 0_u32
+
+    @endpoint : String
+    @server_id : UInt64
+    @session_id : String
+    @token : String
+
+    @heartbeat_interval : Int32?
+
+    # Creates a new voice client. The *payload* should be a payload received
+    # from Discord as part of a VOICE_SERVER_UPDATE dispatch, received after
+    # sending a voice state update (gateway op 4) packet. The *session* should
+    # be the session currently in use by the gateway client on which the
+    # aforementioned dispatch was received, and the *user_id* should be the
+    # user ID of the account on which the voice client is created. (It is
+    # received as part of the gateway READY dispatch, for example)
+    def initialize(payload : Discord::Gateway::VoiceServerUpdatePayload,
+                   session : Discord::Gateway::Session, @user_id : UInt64)
+      @endpoint = payload.endpoint.gsub(":80", "")
```

Comment:
I can't use the URI parser because the endpoint doesn't have a scheme (it's usually just `eu-west1.discord.gg:80` or something)